AWS-SDK Migration

benzekrimaha opened the pull request:
Issue: ARSN-514

Please note that a ticket has been creaed to add coverage not to make this PR more extended : https://scality.atlassian.net/browse/ARSN-524

bert-e commented:
# Hello benzekrimaha,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

SylvainSenechal commented:
> Issue: [ARSN-514](https://scality.atlassian.net/browse/ARSN-514)
> 
> Please note that a ticket has been creaed to add coverage not to make this PR more extended : https://scality.atlassian.net/browse/ARSN-524

Wait, you say you've got two Jira ticket 514/524, but they both point to this same PR 🤔 

[ARSN-514]: https://scality.atlassian.net/browse/ARSN-514?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

SylvainSenechal commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -45,42 +44,25 @@ interface ClientOptions {
             endpoint,
-            httpOptions,
-            ...credentials,
+            requestHandler: undefined, // v3 handles agents differently
```
Aren't we dropping tls support here ? If you say it is still handled but differently, maybe reference where/how it is done. Or maybe it's just not needed ? 
In Vault I was able to use it like this https://github.com/scality/vault2/pull/177/files#diff-a78b27e0448a01885b6f5220e7a97d50c0729ff792356c653a09bf67361446e9R154

Other example in s3utils : https://github.com/scality/s3utils/blob/development/1.16/VerifyReplication/storage/s3.js#L63

SylvainSenechal commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -163,30 +135,23 @@ export default class Client implements KMSInterface {
-
             cb(null);
+        }).catch(err => {
+            if (err.name === 'NotFoundException' || err.name === 'KMSInvalidStateException') {
```
nit: I think it's slightly nicer to use instanceof https://aws.amazon.com/blogs/developer/service-error-handling-modular-aws-sdk-js/

```suggestion
            if (err instanceof NotFoundException || err instanceof KMSInvalidStateException) {
```

francoisferrand replied:
nicer and more importantly safer & recommended approach in SDK docs

SylvainSenechal commented on lib/network/utils.ts:
```diff
@@ -32,6 +31,15 @@ export function arsenalErrorKMIP(err: string | Error) {
 const allowedKmsErrorCodes = Object.keys(allowedKmsErrors) as unknown as (keyof typeof allowedKmsErrors)[];
 
+// Local AWSError type for compatibility with v3 error handling
+export type AWSError = Error & { 
```
I haven't tested/run the code but just wanna double check, you sure this file is ok this way ? I mean specifically, I'm looking at line 50, should err.code be replaced with err.name ?

SylvainSenechal commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -163,30 +135,23 @@ export default class Client implements KMSInterface {
-                return;
-            }
-
+        this.client.send(new ScheduleKeyDeletionCommand(params)).then(data => {
```
nit: I think it's nicer to always have client.send(command) and define command earlier, instead of having params as an untyped struct
```suggestion
        const command = new ScheduleKeyDeletionCommand({
            ....
        })
        this.client.send(command).then(data => {
```

SylvainSenechal replied:
Yeah actually just like you did in the healthcheck() function at the end of this file

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -63,9 +63,18 @@ class AwsClient {
                 return cb();
+            })
+            .catch(err => {
+                if (err && err.code !== 'AuthorizationHeaderMalformed') {
```
Did you double check all the err.code vs err.name ?

SylvainSenechal replied:
Also, I think no need to check "if err", it was needed previously because we were receiving both (err, res) in the same scope

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -27,33 +40,20 @@ class AwsClient {
                     region = res.LocationConstraint;
                 }
-                this._client.config.update({ region });
+                this._client = new S3Client({ ...this._s3Params, region });
```
The old code is a little weird imo, (with region being set to err.region in case of an error??), but I feel like here, if region is not defined, we should just return without changing anything to the client. Because we can end up in a weird situation where we update the client region, but not the client endpoint (which is probably not wanted?) 
Also de we need to create a new client, or can we just update the existing client the way it was done previously ?

benzekrimaha replied:
With the new SDK we indeed can't update the client config , we need to create a new one with the new config ( aka the new region here )

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -116,7 +125,8 @@ class AwsClient {
         if (keyContext.isDeleteMarker) {
-            return this._client.deleteObject(params, putCb);
+            return this._client.send(new DeleteObjectCommand(params)).then(res => putCb(null, res))
+            .catch(err => putCb(err));
```
```suggestion
            .catch(putCb);
```
Can double check other occurrences in this file & elsewhere

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -168,50 +180,35 @@ class AwsClient {
+            };
+            return callback(null, response);
+        }).catch(err => {
+            if (err.code === 'NoSuchKey' || err.code === 'NotFound') {
```
I don't wanna leave the same comment on all occurence of this err.code vs err.name situation, but make sure this is ok. Because in the Zenko migration, I just checked and when I used getObjectCommand, i checked the err with err.name : https://github.com/scality/Zenko/blob/improvement/ZENKO-5054/tests/zenko_tests/node_tests/cloudserver/keyFormatVersion/tests/versionedBucket.js#L232

SylvainSenechal replied:
<img width="591" height="414" alt="image" src="https://github.com/user-attachments/assets/b54b0e83-199e-4e37-b8c7-f32142b6eafa" />

SylvainSenechal replied:
@benzekrimaha double checking after your changes : 
I see you used xxxException, but I can't import them, instead I can import xxx without the 'Exception' as you can see for example in the screenshot : NoSuchKey exists, but NoSuchKeyException doesn't exists.
Not too sure if it's me who has a messed up IDE or something, but worth double checking 

<img width="280" height="93" alt="image" src="https://github.com/user-attachments/assets/c9d2fd80-86cb-4afa-b59b-4b781436a0a4" />

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -237,44 +235,39 @@ class AwsClient {
-                    Bucket: this._awsBucketName },
-                (err, data) => {
-                    if (err) {
+                return this._client.send(new GetObjectCommand({
```
Is there a reason you didn't use GetBucketVersioningCommand ?

SylvainSenechal commented on tests/unit/storage/data/locConstraintParser.spec.js:
```diff
@@ -38,22 +38,26 @@ describe('locationConstraintParser', () => {
+            assert.strictEqual(client._s3Params.httpOptions.agent.protocol, 'https:');
+            assert.strictEqual(client._s3Params.httpOptions.agent.keepAlive, false);
+        }
+        // signatureVersion is not used in AWS SDK v3, skip this assertion
```
I think can just delete these 🤔  Or indicate this was used in V2

SylvainSenechal commented on tests/unit/storage/data/DummyService.js:
```diff
@@ -178,7 +192,117 @@ class DummyService {
     }
-    // To-Do: add tests for other methods
+    
+    send(command) {  // Remove callback parameter completely
```
This comment is a bit confusing, I feel like it's more aimed for code reviewers and shouldnt stay in the file 🤔

SylvainSenechal commented on tests/functional/kmsAWS/highlevel.spec.js:
```diff
@@ -26,46 +21,74 @@ describe('KmsAWSClient', () => {
+        });
     });
 
     it('should create a new master encryption key', done => {
```
Looks like a duplicate of the first test "should support default encryption key per account" 🤔

SylvainSenechal replied:
Up, still here

SylvainSenechal commented on package.json:
```diff
@@ -25,7 +25,9 @@
     "ajv": "6.12.3",
     "async": "~2.6.4",
-    "aws-sdk": "^2.1691.0",
+    "@aws-sdk/client-s3": "^3.540.0",
```
I checked the 3.540 release, it's already from Mar 22, 2024. Do you think we can pick something more recent ? I guess it should be compatible with current migrated code

SylvainSenechal commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -0,0 +1,538 @@
+                .customizeDescription('You must specify at least one part');
+            return callback(error);
+        }
+        for (let ind = 1; ind < partList.length; ++ind) {
```
Not sure if this is likely to happen in practice, but does that need stricter checking than simple ordering ? For example here, [1,2,10] would be valid here but maybe we want to enforce [1,2,3] (or [0,1,2]) ?
Maybe checking that that partList[i] = i + 1 (or i)

benzekrimaha commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,21 +150,23 @@ class AwsClient {
 
         uploadParams.Body = stream;
-        return this._client.upload(uploadParams, putCb);
+        return this._client.send(new PutObjectCommand(uploadParams)).then(res => putCb(null, res))
```
for the upload , I'm creating another ticket here : ARSN-526 to do so otherwise this PR will be a lot more extensive

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/ARSN-514` and the
destination branch `development/8.3`.

Please resolve the conflict on **the feature branch** (`improvement/ARSN-514`).

```sh
git fetch && \
git checkout origin/improvement/ARSN-514 && \
git merge origin/development/8.3
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/ARSN-514
```

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -168,50 +185,58 @@ class AwsClient {
-            let logLevel;
-            if (err.code === 'NotFound') {
-                logLevel = 'info';
+        this._client.send(new GetObjectCommand(params)).then(data => {
+            // Always return an object with .createReadStream for test compatibility
+            const stream = data.Body;
+            const abort = () => {
+                if (isAborted) return; // Prevent multiple abort calls
+                isAborted = true;
+                
```
I'm not sure to see where this variable is defined? We first initialize it in the next line, shouldn't we create it out of the scope of `abort`?

```suggestion
        this._client.send(new GetObjectCommand(params)).then(data => {
            // Always return an object with .createReadStream for test compatibility
            const stream = data.Body;
            let isAborted = false;
            const abort = () => {
                if (isAborted) {
                    return; // Prevent multiple abort calls
                }
                isAborted = true;
                
```

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -168,50 +185,58 @@ class AwsClient {
+                isAborted = true;
+                
+                if (stream) {
+                    if (typeof stream.destroy === 'function') {
+                        stream.destroy();
+                    } else if (typeof stream.abort === 'function') {
+                        stream.abort();
+                    } else if (typeof stream.close === 'function') {
+                        stream.close();
+                    } else if (typeof stream.end === 'function') {
+                        stream.end();
+                    }
```
```suggestion
                    (stream?.destroy || stream?.abort || stream?.close || stream?.end)?.();
```

?

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -168,50 +185,58 @@ class AwsClient {
+                        stream.end();
+                    }
+                    
+                    if (typeof stream.removeAllListeners === 'function') {
+                        stream.removeAllListeners();
+                    }
```
```suggestion
                    stream.removeAllListeners?.();
```

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -168,50 +185,58 @@ class AwsClient {
+                    }
+                }
+            };
+            if (stream && !stream.abort) {
+            stream.abort = abort;
+            }
```
```suggestion
            if (!stream?.abort) {
                stream.abort = abort;
            }
```

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -168,50 +185,58 @@ class AwsClient {
+            if (stream && !stream.abort) {
+            stream.abort = abort;
+            }
+            const response = {
+                createReadStream: () => stream,
+                Body: stream,
+                abort,
+                destroy: abort,
+            };
+            return callback(null, response);
```
```suggestion
            return callback(null, {
                createReadStream: () => stream,
                Body: stream,
                abort,
                destroy,
            });
```

(with the new `abort` variable name)

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -168,50 +185,58 @@ class AwsClient {
+        this._client.send(new GetObjectCommand(params)).then(data => {
+            // Always return an object with .createReadStream for test compatibility
+            const stream = data.Body;
+            const abort = () => {
```
```suggestion
            const destroy = () => {
```

As it's basically destroying the stream, and remove all events

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -223,12 +248,14 @@ class AwsClient {
-                'datastore', err, this._dataStoreName, this.clientType);
-                if (err.code === 'NoSuchVersion' || err.code === 'NoSuchKey') {
-                    // data may have been deleted directly from the AWS backend
+        return this._client.send(new DeleteObjectCommand(params)).then(() => {
+            return callback();
+        }).catch(err => {
```
```suggestion
        return this._client.send(new DeleteObjectCommand(params)).then(() => callback()).catch(err => {
```

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -223,12 +248,14 @@ class AwsClient {
+        return this._client.send(new DeleteObjectCommand(params)).then(() => {
+            return callback();
+        }).catch(err => {
+            logHelper(log, 'error', 'error deleting object from ' +
+            'datastore', err, this._dataStoreName, this.clientType);
```
```suggestion
            logHelper(log, 'error', 'error deleting object from ' +
                'datastore', err, this._dataStoreName, this.clientType);
```

williamlardier commented on lib/storage/data/external/GCP/GcpClient.js:
Currently I see two files named the same, 
`lib/storage/data/external/GcpClient.js`

and

`lib/storage/data/external/GCP/GcpClient.js`

I'm not sure this is expected, or maybe we can find another name for this one?

williamlardier commented on package.json:
```diff
@@ -17,6 +17,10 @@
   "homepage": "https://github.com/scality/Arsenal#readme",
   "dependencies": {
+    "@aws-sdk/client-kms": "^3.540.0",
+    "@aws-sdk/client-s3": "^3.540.0",
```
```suggestion
    "@aws-sdk/client-s3": "^3.876.0",
```

williamlardier commented on lib/storage/data/external/GcpClient.js:
```diff
@@ -13,7 +13,8 @@ const { ListObjectsCommand,
     CopyObjectCommand,
-    HeadObjectCommand } = require('@aws-sdk/client-s3');
+    HeadObjectCommand,
+    AccessDenied } = require('@aws-sdk/client-s3');
```
Errors are not exposed by the lib, from what I can see `AccessDenied` would here always be undefined
It should be `AccessDeniedException` (cf https://aws.amazon.com/blogs/developer/service-error-handling-modular-aws-sdk-js/)

williamlardier commented on lib/storage/data/LocationConstraintParser.js:
```diff
@@ -1,6 +1,7 @@
 const { http, https } = require('httpagent');
 const url = require('url');
 const { S3Client } = require('@aws-sdk/client-s3');
+const { NodeHttpHandler } = require('@aws-sdk/node-http-handler');
```
I think we want to use this one, that is the most up to date version (the aws sdk one is not maintained anymore it seems)

https://www.npmjs.com/package/@smithy/node-http-handler

williamlardier commented on lib/storage/data/LocationConstraintParser.js:
```diff
@@ -76,19 +77,26 @@ function parseLC(config, vault) {
                 endpoint: `${protocol}://${endpoint}`,
                 region,
+                requestHandler: new NodeHttpHandler({
+                    socketTimeout: 0,
```
From this doc it seems we should use `requestTimeout`

https://github.com/aws/aws-sdk-js-v3/blob/main/supplemental-docs/CLIENTS.md#request-handler-requesthandler

williamlardier commented on lib/storage/data/external/GCP/GcpClient.js:
```diff
@@ -0,0 +1,522 @@
+        }))
+            .then(() => callback(null, {}))
+            .catch(err => {
+                // If error is NoSuchKey, bucket exists but key doesn't (which is expected)
+                if (err instanceof NoSuchKey || err instanceof NotFound) {
+                    callback(null, {});
+                }
+                // Other errors likely mean bucket doesn't exist or access denied
+                callback(err);
```
otherwise we call both callbacks

```suggestion
                // If error is NoSuchKey, bucket exists but key doesn't (which is expected)
                if (err instanceof NoSuchKey || err instanceof NotFound) {
                    return callback(null, {});
                }
                // Other errors likely mean bucket doesn't exist or access denied
                return callback(err);
```

williamlardier commented on lib/storage/data/external/GCP/GcpClient.js:
```diff
@@ -0,0 +1,522 @@
+            fakeRequest.gotBucketNameFromHost = true;
+        }
+        const data = {};
+        const logger = { trace: () => {} };
+        const stringToSign = constructStringToSignV2(fakeRequest, data, logger, 'GCP');
```
I think we should probably use a real class logger here, and properly pass the request IDs, as here it would remove any logs in case of real issues?
(can be in a followup ticket?)

williamlardier replied:
up on this, just so we don't forget about the ticket

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -237,44 +264,39 @@ class AwsClient {
                     awsResp[location] = { error: err, external: true };
                     return callback(null, awsResp);
-                }
+        }).then(() => {
```
Maybe not the order of promise we want, as after calling the callback in the .catch, we will execute the .then?

williamlardier commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -427,21 +446,20 @@ class AwsClient {
                   `${this.type}: ${err.message}`),
                     );
-                }
+                }).then(completeMpuRes => {
```
Better to keep the .then before the .catch here I think, same reason as above comment

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -63,9 +68,18 @@ class AwsClient {
                         endpoint,
                     });
-                    this._client.endpoint = new AWS.Endpoint(endpoint);
+                    this._client = new S3Client({ ...this._s3Params, endpoint });
```
I think it needs region ?
```suggestion
                    this._client = new S3Client({ ...this._s3Params, region, endpoint });
```

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -27,33 +45,20 @@ class AwsClient {
-                // set regional endpoint
-                    region = err.region;
-                } else if (res) {
+                if (res) {
```
Pretty sure if (res) is not needed, but LocationConstraint can be undefined according to the typescript structure, should be fine though, if the region is undefined new S3Client handles it

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -168,50 +185,43 @@ class AwsClient {
+                (stream?.destroy || stream?.abort || stream?.close || stream?.end || stream?.removeAllListeners)?.();
+            };
+            if (!stream?.abort) {
+                stream.abort = abort;
```
Wait, abort is not defined, should it be destroy ?

SylvainSenechal replied:
Yeah after reviewing William's comment (https://github.com/scality/Arsenal/pull/2482#discussion_r2397467103), I think you changed the function name was changed but this line wasn't changed

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/ARSN-514` and the
destination branch `development/8.3`.

Please resolve the conflict on **the feature branch** (`improvement/ARSN-514`).

```sh
git fetch && \
git checkout origin/improvement/ARSN-514 && \
git merge origin/development/8.3
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/ARSN-514
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

DarkIsDude commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -45,42 +51,25 @@ interface ClientOptions {
             endpoint,
-            httpOptions,
-            ...credentials,
+            requestHandler: undefined,
```
Why it's needed ?

DarkIsDude commented on lib/network/utils.ts:
```diff
@@ -32,20 +31,37 @@ export function arsenalErrorKMIP(err: string | Error) {
 
 const allowedKmsErrorCodes = Object.keys(allowedKmsErrors) as unknown as (keyof typeof allowedKmsErrors)[];
 
+// Local AWSError type for compatibility with v3 error handling
```
I'm not sure to understand the goal of the comment here, can you explain 🙏

benzekrimaha replied:
before we used to rely on the old aws sdk library which is no longer the case with the migration so had to define the type in order to keep the compatibility with the error handling

DarkIsDude replied:
Then maybe just a naming like `ArsenalAWSError` ?

benzekrimaha replied:
keeping the same name avoids additionnal not needed changes as the PR is already extensive

DarkIsDude commented on lib/storage/data/external/GcpClient.js:
```diff
@@ -298,6 +304,42 @@ class GcpClient extends AwsClient {
+    listObjects(params, callback) {
+        return this.send(new ListObjectsCommand(params))
+            .then(data => callback && callback(null, data))
+            .catch(err => callback && callback(err));
```
```suggestion
            .catch(err => callback?.(err));
```

DarkIsDude commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -86,73 +87,123 @@ const operations = [
     },
 ];
 
+// Reusable cleanup function
```
```suggestion
```

DarkIsDude commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -86,73 +87,123 @@ const operations = [
+                endpoint: `http://${host}`,
+                maxAttempts: 1,
+                forcePathStyle: false,
+                region: 'us-east-1', // Required by AWS SDK v3
```
```suggestion
                region: 'us-east-1',
```

This comment make sense because you are switching the AWS SDK but in 3 months, this comment will not make sense anymore

DarkIsDude commented on tests/unit/storage/data/DummyService.js:
```diff
@@ -178,7 +189,116 @@ class DummyService {
+    
+    send(command) {
+        if (command instanceof PutObjectCommand) {
+            return new Promise((resolve, reject) => {
```
Maybe the new Promise can be moved before all if ? Also, does it make sense to create a switch ?

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -1,4 +1,22 @@
+    GetBucketVersioningCommand,
+    NoSuchKey,
+    NotFound,
+    NoSuchVersion,
```
I'm not too sure NotFound and NoSuchVersion exists : I can't import them 

<img width="476" height="172" alt="image" src="https://github.com/user-attachments/assets/ccf6bf1f-2398-4438-955c-fee09b03ae3b" />

benzekrimaha replied:
Good catch , I assumed it was there as I checked for the noSuchKey but we don't have noSuchVersion , switching back on checking with the err.name for this case ( same for the AccessDeniedException I was relying on as well )
<img width="568" height="406" alt="image" src="https://github.com/user-attachments/assets/e4dbcc66-97d3-4990-a606-27ce31cbc6fe" />

SylvainSenechal commented on lib/storage/data/external/AzureClient.js:
```diff
@@ -263,7 +263,7 @@ class AzureClient {
                 let logLevel;
                 let retError;
-                if (err.code === 'NotFound') {
+                if (err.name === 'NotFound') {
```
I think err.name is ok, could be instanceof NotFound too

SylvainSenechal commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -168,50 +183,44 @@ class AwsClient {
-                logLevel = 'info';
+        this._client.send(new GetObjectCommand(params)).then(data => {
+            // Always return an object with .createReadStream for test compatibility
+            const stream = data.Body;
```
Side note about readStream, sorry the comment is coming so late but I only realized this when working on the new backbeat client : The data body should have these 3 methods to deal with the stream : 
<img width="568" height="296" alt="image" src="https://github.com/user-attachments/assets/5922c499-0166-42a8-ab55-92177c43d1a5" />

This screenshot is for a getObject in the Smithy client I created, but even on the AWS doc, I can find these functions : https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/s3/command/GetObjectCommand/

My point is, I'm not too sure if we should go with your custom solution, or could data.Body.transformToString() be enough ?

benzekrimaha replied:
Actually several tests rely on the .createReadStream function this is why I am introducing it with this change in order to keep compatibility , in arsenal we don't ( in the tests ) actually work with the body thus why I didn't use the transform ( you'll see it used a lot on cloudserver ) , I am introducing it in the response in order to not have to do extensive changes on the tests that's the whole purpose

SylvainSenechal commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,538 @@
+    PutObjectTaggingCommand,
+    GetObjectTaggingCommand,
+    DeleteObjectTaggingCommand,
+    ListPartsCommand } = require('@aws-sdk/client-s3');
```
```suggestion
    ListPartsCommand,
} = require('@aws-sdk/client-s3');
```

SylvainSenechal commented on lib/storage/data/external/GCP/GcpService.js:
Don't know if it requires changes, but it's not clear to me what is the difference between external/GcpClient.js and external/GCP/GcpService.js
They both look like Gcp clients, but GcpService implements more functions, and they extends different classes S3Client vs AwsClient

benzekrimaha replied:
Indeed as before we were relying on the service definition like in bb client , we were using the model.json to define the extended s3client method , and this is what gcp service does here while keeping the old client that extends the awsclient => we have missing method in awsclient that are in the s3client thus the reason behind me keeping both otherwise I'd have to actually redefine the method withing gcpClient. I went for this in order to keep the code easy to read/understand

williamlardier commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -163,30 +160,24 @@ export default class Client implements KMSInterface {
+        }).catch(err => {
+            if (err instanceof NotFoundException || err instanceof KMSInvalidStateException) {
+                logger.info('AWS KMS: key does not exist or is already pending deletion', { masterKeyId, error: err });
+                return cb(null);
```
If we return no error we want at least to have a warning log I think?

```suggestion
                logger.warn('AWS KMS: key does not exist or is already pending deletion', { masterKeyId, error: err });
                return cb(null);
```

williamlardier commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -199,31 +190,24 @@ export default class Client implements KMSInterface {
         const params = {
             KeyId: masterKeyId,
-            KeySpec: 'AES_256',
+            KeySpec: 'AES_256' as const,
```
```suggestion
            KeySpec: 'AES_256',
```

benzekrimaha replied:
needs to be a const otherwise won't match the generateDataKey params typing

williamlardier commented on lib/network/utils.ts:
```diff
@@ -32,20 +31,37 @@ export function arsenalErrorKMIP(err: string | Error) {
 const allowedKmsErrorCodes = Object.keys(allowedKmsErrors) as unknown as (keyof typeof allowedKmsErrors)[];
 
+// Local AWSError type for compatibility with v3 error handling
+export type AWSError = Error & {
```
I'm pretty sure you can remove this and use the `S3ServiceException` type, this is exposed by the `@aws-sdk/client-s3` module and all exceptions related to S3 are extending this base type, it has all the fields you have here, with the advantage of being typed & documented

williamlardier commented on tests/unit/storage/data/external/ExternalClients.spec.js:
```diff
@@ -83,82 +92,99 @@ describe('external backend clients', () => {
                 const uploadId = 'externalBackendTestUploadId';
                 testClient.completeMPU(jsonList, null, key,
                     uploadId, bucketName, log, (err, res) => {
+                        if (err) return done(err);
```
```suggestion
                        if (err) {
                            return done(err);
                        }
```

williamlardier commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,539 @@
-const AWS = require('aws-sdk');
-const Service = AWS.Service;
-
+// GcpClient.js: GCP-compatible S3 client using AWS SDK v3
```
```suggestion
```

(already documented in the class)

williamlardier commented on lib/storage/data/external/GcpClient.js:
```diff
@@ -298,6 +304,42 @@ class GcpClient extends AwsClient {
+
+    listObjects(params, callback) {
+        return this.send(new ListObjectsCommand(params))
+            .then(data => callback && callback(null, data))
```
```suggestion
            .then(data => callback?.(null, data))
```

williamlardier reviewed (approved):
mostly code style, so approving, could be nice to test if we can use `S3ServiceException`, it would allow better error handling logic I think

DarkIsDude commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -45,42 +54,39 @@ interface ClientOptions {
 
     constructor(options: ClientOptions) {
         this._supportsDefaultKeyPerAccount = true;
-        const { providerName, tls, ak, sk, region, endpoint, noAwsArn } = options.kmsAWS;
```
lint : nit

SylvainSenechal replied:
up, can fix it if you make more changes

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-514 contains:


* `8.3.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.3.0`

* `8.4.0`


Please check the `Fix Version/s` of ARSN-514, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

benzekrimaha committed:
GCP client migration

As we can no longer use the Service.defineService
previously used in aws-sdk v2, we have to manually
this commit aims to implement the GCP client
using the new aws-sdk v3 architecture
and migrate the existing code to use it.
Please note that now both the setup and signing
mechanisms had to be re-implemented in the GcpService
file.
Issue: ARSN-514

benzekrimaha committed:
Gcp MPU helper migration

Issue: ARSN-514

benzekrimaha committed:
kms client migration

Issue: ARSN-514

benzekrimaha committed:
aws client migration

In this commit several changes were done as aws-sdk v3
happens to be less permissive than v2 in terms of error handling and
stream management. To ensure backward compatibility with existing
functionalities adjustments were made, please refer to
code comments for more details.
Issue: ARSN-514

benzekrimaha committed:
data related files migration

Issue: ARSN-514

benzekrimaha committed:
auth related files migration

Issue: ARSN-514

benzekrimaha committed:
utils files migration

Issue: ARSN-514

benzekrimaha committed:
functional tests migration

Issue: ARSN-514

benzekrimaha committed:
unit tests migration

Issue: ARSN-514

benzekrimaha committed:
Azure client migration adaptation

As the sdk now returns name instead of code for errors,
we need to adapt the error handling accordingly.
Issue: ARSN-514

benzekrimaha committed:
deps install

A security issue was fixed using resolutions in package.json
as the vulnerability is in a transient dependency.
Issue: ARSN-514

SylvainSenechal commented:
There is still a double test here, or am I miss reading ? 🤔  https://github.com/scality/Arsenal/pull/2482#discussion_r2367443617

SylvainSenechal commented on lib/auth/v4/streamingV4/constructChunkStringToSign.ts:
```diff
@@ -24,9 +24,7 @@ export default function constructChunkStringToSign(
-        const temp = justDataChunk instanceof Buffer
-            ? hash.update(justDataChunk)
-            : hash.update(justDataChunk, 'binary');
+        const temp = hash.update(justDataChunk);
```
I don't have a lot of opinion about this, but what about William's comment ? is it ok to remove the 'binary' option ?

SylvainSenechal commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -118,31 +100,21 @@ export default class Client implements KMSInterface {
-                // Prefer ARN, but fall back to KeyId if ARN is missing
-                // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
-                keyId = keyMetadata?.Arn ?? keyMetadata?.KeyId!;
+                keyId = keyMetadata?.Arn ?? (keyMetadata?.KeyId || '');
```
Probably still applicable

SylvainSenechal commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -300,41 +274,30 @@ export default class Client implements KMSInterface {
 
     /**
-     * NOTE1: S3C-4833 KMS healthcheck is disabled in CloudServer
-     * NOTE2: The best approach for implementing the AWS KMS health check is still under consideration.
```
This is not relevant anymore, right ?

benzekrimaha replied:
it has been removed in my PR as I am enabling it

francoisferrand replied:
is there an impact (on S3C) of enabling this?
typically better to do in a separate PR, to avoid scope creeping and let knowledgeable
people to jump in...

SylvainSenechal commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -22,25 +34,86 @@ function handler(isPathStyle) {
 }
 
 const invalidDnsBucketNames = [
-    '..',
```
could've keep this one no ?

SylvainSenechal commented:
Only nits from me. I think it would be nice to have a review from François though @benzekrimaha

benzekrimaha committed:
fixups post reviews

francoisferrand commented on lib/auth/v2/getCanonicalizedAmzHeaders.ts:
```diff
@@ -9,7 +9,16 @@ export default function getCanonicalizedAmzHeaders(headers: Record<string, strin
-        .map(val => [val.trim(), headers[val].trim()]);
+        .map(val => {
+            const headerValue = headers[val];
+            // AWS SDK v3 can pass header values as arrays (for multiple values),
```
if this is the case, then the `getCanonicalizedAmzHeaders()` does not have the right type: headers
should be of type `ArsenalRequestHeaders`...

(and i find it a bit worrying that typescript compiler did not complain...)

benzekrimaha replied:
in the ArsenalRequestHeaders we have the string[] , the cases I see on our tests in the other hand are sometimes passing numbers ( hasn't been introduced by the migration ), as for the compiler complaining I think it's because the tests are in spec.js extensions. That comment emphasizes we’re just normalising anything supplied by tests, the real sdk only produces strings

francoisferrand replied:
If "the real sdk only produces strings", then
- the test should pass only strings as well
- the comment give a false indication that "AWS SDK v3 can pass header values as arrays"

And again, the whole point of typescript is to improve the stability of the code by properly typing. So either
- the comment is wrong, we can only receive string : the typing is correct but comment needs to be fixed, maybe some tests as well, and this chunk doing the normalization must be dropped
- or the SDK can indeed send a string array: we must keep the normalization indeed, but the type of the function must be updated to match reality.

francoisferrand commented on lib/auth/v4/streamingV4/constructChunkStringToSign.ts:
```diff
@@ -24,10 +24,12 @@ export default function constructChunkStringToSign(
-            ? hash.update(justDataChunk)
-            : hash.update(justDataChunk, 'binary');
-        currentChunkHash = temp.digest('hex');
+        if (typeof justDataChunk === 'string') {
```
why this change?

Looking at the signature of the function, this is exactly the same: either this change is not needed, or the function signature should be updated...

francoisferrand commented on lib/auth/v4/createCanonicalRequest.ts:
```diff
@@ -73,7 +73,14 @@ export default function createCanonicalRequest(
+            // AWS SDK v3 can pass header values as arrays (for multiple values),
+            // strings, or other types. We need to normalize them before calling .trim()
+            // Per HTTP spec and AWS Signature v4, multiple values are joined with commas
+            const stringValue = Array.isArray(headerValue) 
```
if we have multiple such places, should we not introduce a helper function?
(may help to type the function, i.e. finding out what the actual type of pHeaders is: is it
ArsenalHeaders ?)

francoisferrand commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -118,31 +124,26 @@ export default class Client implements KMSInterface {
-                // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
-                keyId = keyMetadata?.KeyId!;
+                keyId = keyMetadata?.KeyId || '';
             } else {
                 // Prefer ARN, but fall back to KeyId if ARN is missing
-                // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
-                keyId = keyMetadata?.Arn ?? keyMetadata?.KeyId!;
+                keyId = keyMetadata?.Arn ?? (keyMetadata?.KeyId || '');
             }
```
why those changes, does it really happen since the SDK upgrade?

The eslint/typescript warning is "free", why this change add extra code (and load) on the path: fine
if needed, but if it actually cannot happen best to keep as is for now and eventually fix the typing
so we don't need coercion anymore (without introducing fallback)...

francoisferrand replied:
ping?

francoisferrand commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -163,30 +164,24 @@ export default class Client implements KMSInterface {
             cb(null);
+        }).catch(err => {
+            if (err instanceof NotFoundException || err instanceof KMSInvalidStateException) {
+                logger.info('AWS KMS: key does not exist or is already pending deletion', { masterKeyId, error: err });
```
keep comment and wrap line (as before)

```suggestion
                // master key does not exist or is already pending deletion
                logger.info('AWS KMS: key does not exist or is already pending deletion',
                    { masterKeyId, error: err });
```

francoisferrand commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -262,6 +243,10 @@ export default class Client implements KMSInterface {
             logger.debug('AWS KMS: data key ciphered');
             cb(null, Buffer.from(data.CiphertextBlob as Uint8Array));
             return;
```
nit: might as well remove the `return;` as well

francoisferrand commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -199,31 +194,24 @@ export default class Client implements KMSInterface {
-
             logger.debug('AWS KMS: data key generated');
             cb(null, isolatedPlaintext, Buffer.from(data.CiphertextBlob as Uint8Array));
+        }).catch(err => {
```
is typescript able to "understand" the type of error automatically?
if not, we should still keep the type in prototype...

```suggestion
        }).catch(err: AWSError => {
```
or (if there is no detail on type, in AWS)
```suggestion
        }).catch(err: error => {
```

(just posting once, applies in multiple places...)

francoisferrand commented on lib/network/utils.ts:
```diff
@@ -32,20 +33,23 @@ export function arsenalErrorKMIP(err: string | Error) {
+function isAWSError(err: unknown): 
+    err is S3ServiceException | KMSServiceException | (Error & { name?: string }) {
+    return (err instanceof S3ServiceException || err instanceof KMSServiceException ||
+        (err instanceof Error && typeof err.name === 'string')
```
this condition seems quite broad, and in particular will be true for every `Error`: since `name` is
a property of the standard javascript `Error` type (and a string by default).

is there a more robust way to identify AWS errors?
or just rename this function, which does not identify AWS errors?
or even drop it altogether and adjust `arsenalErrorAWSKMS` accordingly: in the current code, it
looks like there are 3 cases, though actually we will never go through the last
`_normalizeArsenalError(err, 'AWS_KMS:')` path....

(maybe reviewing the places where `arsenalErrorAWSKMS` is called -and possibly typing- would help to
discard some impossible case or identify "when" we need to go through specific path)

francoisferrand commented on lib/s3middleware/objectUtils.ts:
```diff
@@ -1,13 +1,13 @@
 const msInOneDay = 24 * 60 * 60 * 1000; // Milliseconds in a day.
 
-export const getMD5Buffer = (base64MD5: WithImplicitCoercion<string> | Uint8Array) =>
+export const getMD5Buffer = (base64MD5: string | Buffer | Uint8Array) =>
```
why do we need this change? Does `Buffer` to "implicitely" coerce to string?

benzekrimaha replied:
Fair point—Buffer already fits WithImplicitCoercion<string>, so functionally nothing changes. I expanded the signature so the accepted inputs are obvious to readers who aren’t familiar with that helper type, but I’m fine reverting if we’d rather keep the original declaration.

francoisferrand commented on tests/functional/kmsAWS/highlevel.spec.js:
```diff
@@ -162,114 +201,142 @@ describe('KmsAWSClient', () => {
     it('should handle errors deleting an existing key on bucket deletion', done => {
         const mockError = new Error('mock delete error');
-        scheduleKeyDeletionStub.yields(mockError, null);
+        mockError.name = 'AccessDeniedException';
```
is there no actual "AWS" error type for this?

benzekrimaha replied:
no we don't have it

francoisferrand commented on tests/unit/s3routes/routesUtils/responseStreamData.spec.js:
```diff
@@ -158,36 +159,58 @@ describe('routesUtils.responseStreamData', () => {
+                    try {
+                        const nOpenSockets = Object.keys(awsAgent.sockets).length;
+                        assert.strictEqual(nOpenSockets, 0);
+                        safeDone();
```
this is not "safe" : if it throws, we will call it again in catch...
the proper fix is not to re-implement `jsutils.once()`...

francoisferrand commented on tests/unit/s3routes/routesUtils/responseStreamData.spec.js:
```diff
@@ -158,36 +159,58 @@ describe('routesUtils.responseStreamData', () => {
-            isclosed: true,
-        }, undefined, logger.newRequestLogger());
+    'data backend starts streaming', async () => {
+        await new Promise((resolve, reject) => {
```
no point introducing this Promise, this is pure callback code....

francoisferrand commented on tests/unit/s3routes/routesUtils/responseStreamData.spec.js:
```diff
@@ -158,36 +159,58 @@ describe('routesUtils.responseStreamData', () => {
+                        const nOpenSockets = Object.keys(awsAgent.sockets).length;
+                        assert.strictEqual(nOpenSockets, 0);
+                        safeDone();
+                    } catch (err) {
```
no reason this can throw, the code inside the `try` is really just
`Object.keys(awsAgent.sockets).length` (if it fails, there is a bug in the test) or the assertion
(no point catching it)

francoisferrand commented on tests/unit/storage/data/external/ExternalClients.spec.js:
```diff
@@ -1,6 +1,7 @@
 const assert = require('assert');
 const async = require('async');
```
not used anymore, should be removed

francoisferrand commented on tests/unit/storage/data/external/ExternalClients.spec.js:
```diff
@@ -83,82 +92,90 @@ describe('external backend clients', () => {
+            const readable = result
+            
+            let errorHandled = false;
+            
```
extra spaces & empty lones

francoisferrand commented on tests/unit/storage/data/external/ExternalClients.spec.js:
```diff
@@ -83,82 +92,90 @@ describe('external backend clients', () => {
+            
+            let errorHandled = false;
+            
+            await new Promise((resolve) => {
```
spaces not needed around parameter

```suggestion
            await new Promise(resolve => {
```

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -22,7 +34,66 @@ function handler(isPathStyle) {
+        } else if (req.method === 'GET') {
+            if (req.url.includes('versioning')) {
+                // getBucketVersioning
+                const xml = '<?xml version="1.0" encoding="UTF-8"?><VersioningConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/"/>';
```
wrap long lines (same below)

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -35,12 +106,15 @@ const invalidDnsBucketNames = [
-    res.end();
+function invalidDnsBucketNameHandler() {
+    return (req, res) => {
+        assert(req.headers.host, host);
```
why this indirection?
it makes code harder to read...

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -86,33 +160,68 @@ const operations = [
+            }
+        });
+
+        await new Promise((resolve) => {
```
no parenthesis around single parameter of lamdba

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -86,33 +160,68 @@ const operations = [
+                credentials: {
+                    accessKeyId: accessKeyId,
+                    secretAccessKey: secretAccessKey,
+                },
```
suprises the linter did not see this (it does in my IDE):
```suggestion
                    accessKeyId,
                    secretAccessKey,
                },
```

same a few times below

francoisferrand commented on tests/unit/storage/data/DummyService.js:
```diff
@@ -81,75 +93,161 @@ class AzureDummyContainerClient {
+        this.abortMultipartUploadAsync = promisify(this.abortMultipartUpload.bind(this));
+        
+        this.commandHandlers = new Map([
+            [PutObjectCommand, (cmd) => this.putObjectAsync(cmd.input)],
```
same below:
```suggestion
            [PutObjectCommand, cmd => this.putObjectAsync(cmd.input)],
```

francoisferrand commented on tests/unit/storage/data/locConstraintParser.spec.js:
```diff
@@ -38,22 +38,22 @@ describe('locationConstraintParser', () => {
-        assert.strictEqual(client._s3Params.httpOptions.agent.keepAlive, false);
-        assert.strictEqual(client._s3Params.signatureVersion, 'v4');
+        assert(client._s3Params.endpoint.startsWith('https://'));
+        if (client._s3Params.httpOptions && client._s3Params.httpOptions.agent) {
```
should not have a condition: either this is a HTTP test, or it is an HTTPS one...

francoisferrand commented on tests/unit/storage/data/locConstraintParser.spec.js:
```diff
@@ -38,22 +38,22 @@ describe('locationConstraintParser', () => {
-        assert.strictEqual(client._s3Params.httpOptions.agent.keepAlive, false);
-        assert.strictEqual(client._s3Params.signatureVersion, 'v2');
+        assert(client._s3Params.endpoint.startsWith('http://'));
+        if (client._s3Params.httpOptions && client._s3Params.httpOptions.agent) {
```
same

francoisferrand commented on package.json:
```diff
@@ -117,6 +122,9 @@
       }
     }
   },
+  "resolutions": {
```
why do we need `resolutions`?
isn't thre something else to fix/check instead?

benzekrimaha replied:
no as the tar is not a direct dependency the cleanest way I could find was the resolution here , I can create a ticket to dig deeper in a separate PR.

francoisferrand commented on lib/storage/data/external/GCP/GcpApis/mpuHelper.js:
```diff
@@ -187,6 +187,10 @@ class MpuHelper {
                 };
                 this.service.deleteObject(delParams, err => {
                     if (err) {
+                        if (err.name === 'NoSuchKey') {
```
why do we need to _ignore_ this error in here?
the behavior of the API did not change, so we should not start to ignore the error?

benzekrimaha replied:
Ignoring NoSuchKey here is about matching S3 semantics and keeping the cleanup idempotent, not relaxing validation in the caller-visible API. removeParts is a best-effort garbage-collection step running after we’ve already returned the composed object; when two retrying threads hit the same MPU bucket, one can delete a part before the other, and GC service responds with NoSuchKey (unlike S3, which treats the delete as a success). Letting that bubble up aborts the whole cleanup even though there is nothing left to delete. By swallowing just that code we align with S3’s “deleting a non-existent key succeeds” behaviour and avoid spurious failures caused by eventual-consistency/listed-but-already-removed parts. Any other error still makes us bail out, so external behaviour stays the same while retry/idempotency improves.

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,908 @@
+    PutObjectTaggingCommand,
+    GetObjectTaggingCommand,
+    DeleteObjectTaggingCommand,
+    ListPartsCommand,
```
never used:
```suggestion
```

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,908 @@
+        // but signing libraries expect strings. This middleware normalizes all
+        // header values to strings before signing.
+        this.middlewareStack.add(
+            (next) => async (args) => {
```
```suggestion
            next => async args => {
```

(many places below as well)

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,908 @@
+        // header values to strings before signing.
+        this.middlewareStack.add(
+            (next) => async (args) => {
+                if (args.request && args.request.headers) {
```
this middelware should be extracted in "separate" middleware (not inline in the middle of the code).

(and maybe we can use it as well -or factorize at least- to handle header normalization in our `auth` module?)

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,908 @@
+        const command = new PutObjectCommand(params);
+        
+        // Capture GCP-specific headers
+        let gcpHeaders = {};
```
nit: no point storing evert header : we use only `x-goog-generation`, so best to store this one only and
avoid pontially wasting memory...

```
let gcpGeneration = undefined
command.middlewareStack.add(
    (next) => async (args) => {
        const result = await next(args);
        if (result.response?.headers) {
            gcpGeneration = result.response?.headers['x-goog-generation'];
        }
        return result;
    },
    {
        step: 'deserialize',
        name: 'captureGcpHeaders',
    }
);

return this.send(command).then(data => {
    return callback && callback(null, {
        ...data,
        VersionId: gcpGeneration, // Add GCP-specific fields
    });
}).catch(err => callback?.(err));

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,908 @@
+     * @param {object} params - S3 putObjectTagging params
+     * @param {function} callback
+     */
+    putObjectTagging(params, callback) {
```
what is the point of implementing every API here?

It was not there before, so not needed : if anything, must be a separate improvement/PR. Unless
absolutely _needed_, should stick to the minimum set of required (i.e. pre-existing) APIs...

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,908 @@
+            }, next),
+            (resObj, next) => {
+                const completeMD = Object.assign({}, resObj.Metadata, tagRes);
+                this.copyObject({
```
putObjectTagging should not be implemented with CopyObject : this would actually duplicate the data,
and cause unexpected (service) charges to the customer...

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -1,13 +1,32 @@
+    GetBucketVersioningCommand,
+    NoSuchKey,
+    NotFound } = require('@aws-sdk/client-s3');
+const { HttpRequest } = require('@smithy/protocol-http');
```
not used

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -26,47 +45,82 @@ class AwsClient {
+                region: 'us-east-1',
+            });
+        } catch (err) {
+            this._logger.error('unexpected error during setup', {
```
why would S3Client fail specifically?
generally, best to have a single try/catch instead of litering the code...

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -26,47 +45,82 @@ class AwsClient {
+            }
+            settled = true;
+            callback(err);
+        };
```
* this is actually a reimplementation of `jsutils.once()` : don't do this, just reuse
* why do we need this? Seems like a workaround of an actual issue...

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -26,47 +45,82 @@ class AwsClient {
+
+            try {
+                this._client = new S3Client(newConfig);
+            } catch (err) {
```
not needed, this should be caught by the promise execution (`.then(r => applyRegion(...))`), and thus
end up in the following `.catch()` block where it will be printed...

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -26,47 +45,82 @@ class AwsClient {
-                    region = res.LocationConstraint;
-                }
-                this._client.config.update({ region });
+        const callback = typeof cb === 'function' ? cb : () => {};
```
why would this function be called without callback?
it is not async, so the caller would not be able to know when the setup is done..

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -26,47 +45,82 @@ class AwsClient {
+        };
+        let usEast1Client;
+        try {
+            usEast1Client = new S3Client({
```
why create another client?
there is one already from constructor: `this._client = new S3Client(this._s3Params);`

benzekrimaha replied:
we can no longer update the client like before we need to create a new one each time

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -26,47 +45,82 @@ class AwsClient {
+                method: 'AwsClient.setup',
             });
+            return done(err);
+        });
```
bad indent

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -90,35 +144,52 @@ class AwsClient {
+                } catch (headErr) {
+                        logHelper(log, 'error', 'failed to retrieve version id for data ' +
+                            'backend object', headErr, this._dataStoreName, this.clientType);
+                }
```
bad indent

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -26,47 +45,82 @@ class AwsClient {
-                return cb();
+        const applyRegion = region => {
+            if (!region) {
+                return done();
```
calling `done()` from within this function makes the code very hard to follow, especially as it is
called as from within the `catch()` call.

Probably better to either use `callbackify` (stick to the old code structure, with a single callback
so no duplication/need to introduce a function/...) or fully embrace async and rewrite the logic
accordingly (invoking the callback on exit)

francoisferrand replied:
ping

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -90,35 +144,52 @@ class AwsClient {
+                        Bucket: this._awsBucketName,
+                        Key: awsKey
+                    });
+                    const headResult = await this._client.send(headCommand);
```
this creates an extra request for every Put, this is not acceptable.

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +211,107 @@ class AwsClient {
+            const command = new PutObjectCommand(uploadParams);
+            return this._client.send(command)
+                .then(data => {
+                    putCb(null, data)})
```
lint issue:


```suggestion
                    putCb(null, data)};
                )
```
or
```suggestion
                .then(data => putCb(null, data))
```

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +211,107 @@ class AwsClient {
+        
+        return upload.done()
+            .then(data => {
+                putCb(null, data)})
```
same

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +211,107 @@ class AwsClient {
-                'from datastore', err, this._dataStoreName);
-                return callback(retError);
+        })).then(data => callback(null, data))
+            .catch(err => {
```
wrong indent

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +211,107 @@ class AwsClient {
-            logHelper(log, logLevel,
-                `error streaming data from ${this.type}`,
-                err, this._dataStoreName, this.clientType);
+            let isAborted = false;
```
we use to setup handler to log errors, not there anymore

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +211,107 @@ class AwsClient {
+                    return;
+                }
+                isAborted = true;
+                (stream?.destroy || stream?.abort || stream?.close || stream?.end || stream?.removeAllListeners)?.();
```
?????

benzekrimaha replied:
the method doesn't return a stream anymore , so we need to check this otherwise we end up with failures.

francoisferrand replied:
> the method doesn't return a stream anymore

it does not return a stream, I get that.
But it returns "something", which is of a known type, and which we must handle properly.

There is no way we want to just "try whatever function is present on the object we get" : we need to understand what type is returned, and process it properly as a consequence.

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +211,107 @@ class AwsClient {
+            // AWS SDK v3 returns streams that may not be fully compatible Node.js Readable streams
+            // They might have .pipe() but not .on(), .once(), etc.
+            // Always check if we have a proper Node.js stream with full API
+            const isProperNodeStream = stream && 
```
@SylvainSenechal had similar issue in backbeat, and solved by fully embracing the actual aws sdk v3
api : we should do the same... It's not a "proper" node.js stream anymore, so we have to adapt.

benzekrimaha replied:
I can create a follow up ticket for this not to extend the changes further

francoisferrand replied:
since your change actually works, it means you handle it already:
- in this PR, you must already remove (i.e. not introduce) dead code : i.e. handling everything that is not in the SDK anymore. Whatever we get, we *must* understand what it is and handle this case specifically (and not others)
- maybe there is a better way (not transforming to web stream, not adding an extra createReadStream... and possibly switching to AbortController to handle "interruption"), that may go to a separate PR...

Note though that it may make comparison (with the current "stable" code) more complex, as we may get biases by this first attempt: so best to at least _investigate_ it, to understand what it would mean (and anyway doing the cleaning is needed)

Probably worth doing a peer programming session with @SylvainSenechal, to review (now with all accumulated knowledge) how we can/should rework this specific function (which is not big anyway)

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +211,107 @@ class AwsClient {
+                typeof stream.on === 'function' &&
+                typeof stream.once === 'function';
+            
+            if (!isProperNodeStream) {
```
AWS SDK (v3) will not return a "proper" node.js stream anymore, so no point trying to handle that
case. The only case we need to handle is AWS SDK v3 stream.

benzekrimaha replied:
same as before , will create a follow up ticket for this

francoisferrand replied:
c.f. previous comment, change needed already : we don't want to introduce dead code

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -223,66 +323,56 @@ class AwsClient {
-                    return callback(null, awsResp);
+        this._client.send(new HeadObjectCommand({ Bucket: this._awsBucketName }))
+            .then(() => {
+            if (!this._supportsVersioning) {
```
bad indent

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -303,26 +393,21 @@ class AwsClient {
+        return this._client.send(new CreateMultipartUploadCommand(params)).then(mpuResObj => {
             return callback(null, mpuResObj);
+        }).catch(err => {
+            logHelper(log, 'error', 'err from data backend', 
```
lint
```suggestion
        return this._client.send(new CreateMultipartUploadCommand(params)).then(mpuResObj =>
            callback(null, mpuResObj);
        ).catch(err => {
            logHelper(log, 'error', 'err from data backend', 
```

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -333,51 +418,45 @@ class AwsClient {
+            params.PartNumberMarker = String(partNumberMarker);
+        }
+        return this._client.send(new ListPartsCommand(params))
+        .then(partList => {
```
bad indent (or keep on previous line)

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -427,71 +512,59 @@ class AwsClient {
+            return callback(errorInstances.ServiceUnavailable
+                .customizeDescription('Error returned from ' +
+                `${this.type}: ${err.message}`));
+        })
```
missing semicolon

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -507,20 +580,16 @@ class AwsClient {
+            return callback(errorInstances.ServiceUnavailable
+                .customizeDescription('Error returned from ' +
+                `${this.type}: ${err.message}`));
+        })
```
missing semi colon

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -563,60 +625,55 @@ class AwsClient {
-                );
-            }
+        return this._client.send(new CopyObjectCommand(awsParams))
+        .then(copyResult => {
```
bad indent (or stick on previous line)

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -625,26 +682,26 @@ class AwsClient {
-            }
-            const eTag = removeQuotes(res.CopyPartResult.ETag);
+        return this._client.send(new UploadPartCopyCommand(params))
+        .then(res => {
```
bad indent

francoisferrand commented on lib/storage/data/LocationConstraintParser.js:
```diff
@@ -1,6 +1,8 @@
 const { http, https } = require('httpagent');
 const url = require('url');
-const AWS = require('aws-sdk');
+const { S3Client } = require('@aws-sdk/client-s3');
```
never used

francoisferrand commented on lib/storage/data/external/utils.js:
```diff
@@ -45,7 +45,7 @@ const utils = {
         Object.keys(metaObj).forEach(key => {
             const newKey = key.substring(11);
-            newObj[newKey] = metaObj[key];
+            newObj[newKey] = `${metaObj[key]}`;
```
If we need to convert a string, best to make it explicit:
```suggestion
            newObj[newKey] = string(metaObj[key]);
```

francoisferrand commented on lib/storage/data/external/GcpClient.js:
```diff
@@ -298,6 +304,42 @@ class GcpClient extends AwsClient {
         });
     }
+
+    listObjects(params, callback) {
+        return this.send(new ListObjectsCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    putObject(params, callback) {
+        return this.send(new PutObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    getObject(params, callback) {
+        return this.send(new GetObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    deleteObject(params, callback) {
+        return this.send(new DeleteObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    copyObject(params, callback) {
+        return this.send(new CopyObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    headObject(params, callback) {
+        return this.send(new HeadObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
 }
```
why add these methods?
they should come from AwsClient, as before...

benzekrimaha replied:
this was done to not make more changes , keeping the names we already use in tests. I can create a ticket to actually switch all calls to aws format with the command etc.

benzekrimaha commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -0,0 +1,538 @@
+        return this.listObjects(mpuParams, callback);
+    }
+
+    /**
+     * List objects in a bucket.
+     * @param {object} params - S3 listObjects params
+     * @param {function} callback
+     */
+    listObjects(params, callback) {
```
I would rather do that in a separate PR + we have an async/await dubject being treated so rather keep the support for both and then cleanly move the codebase to async await to avoid extending even more the changes in this one

francoisferrand commented on lib/auth/v4/streamingV4/constructChunkStringToSign.ts:
```diff
@@ -24,10 +24,10 @@ export default function constructChunkStringToSign(
-            ? hash.update(justDataChunk)
-            : hash.update(justDataChunk, 'binary');
-        currentChunkHash = temp.digest('hex');
+        const chunkHash = typeof justDataChunk === 'string'
```
why this change?
is it more efficient?

benzekrimaha replied:
no real efficiency gain here; it’s mostly a type‑safety / clarity change, and either version is fine as long as justDataChunk really is only Buffer | string.

francoisferrand replied:
clarity-wise, it is the same : just reversing the condition.
as far as type-safety, I miss the point: is `typeof` instrinsically safer than `instanceof` ? or is it actually safer to handle anything else (!Buffer and !string) as "plain" instead of "binary" ?

either we can explain why this is better, or we should just keep the old code....

francoisferrand commented on lib/auth/v2/getCanonicalizedAmzHeaders.ts:
```diff
@@ -9,7 +12,14 @@ export default function getCanonicalizedAmzHeaders(headers: Record<string, strin
+            const headerValue = headers[val];
+            // AWS SDK v3 can pass header values as arrays (for multiple values),
+            // strings, or other types. We need to normalize them before calling .trim()
+            // Per HTTP spec and AWS Signature v2, multiple values are joined with commas
```
should this comment not be in `normalizeHeaderValue` instead of being repeated?

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,889 @@
+        
+        // Capture GCP-specific headers
+        let objectGeneration;
+        command.middlewareStack.add(
```
it seems we add this middleware to every command : can't we simply add it to the `client` middleware stack instead?

c.f. https://aws.amazon.com/blogs/developer/middleware-stack-modular-aws-sdk-js/ exemple:

```
const { S3 } = require("@aws-sdk/client-s3");
const client = new S3({ region: "us-west-2" });
// Middleware added to client, applies to all commands.
client.middlewareStack.add(
  (next, context) => async (args) => {
    args.request.headers["x-amz-meta-foo"] = "bar";
    const result = await next(args);
    // result.response contains data returned from next middleware.
    return result;
  },
  {
    step: "build",
    name: "addFooMetadataMiddleware",
    tags: ["METADATA", "FOO"],
  }
);

await client.putObject(params);
```

benzekrimaha replied:
The only middleware that’s truly command‑agnostic (header normalization) is already on the client middleware stack. The ones added to PutObject / CopyObject / GetObject / Head* are tied to those commands’ specific output shapes, so putting them on the client stack would either run unnecessary logic for all commands or require a less type‑safe switch on commandName. I’ve kept them per‑command (and can factor out a small helper to reduce duplication), what do you think @francoisferrand ?

francoisferrand replied:
yes, this logic must be made reusable, no point repeating it.

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,889 @@
+                }
+                return result;
+            },
+            {
```
this is duplicated in every call, if we can't move it to the client's Middlewarestack we must anyway make it into a reusable function

benzekrimaha replied:
same as this => https://github.com/scality/Arsenal/pull/2482#discussion_r2627622078

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,889 @@
+    }
+
+    putBucketVersioning(params, callback) {
+        if (!params || !params.Bucket || !params.VersioningConfiguration) {
```
does the SDK not do such validation already?

benzekrimaha replied:
This explicit check is there to preserve the previous behavior (early failure with our InvalidRequest error) and avoid sending obviously invalid requests

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,889 @@
+    }
+
+    getBucketVersioning(params, callback) {
+        if (!params || !params.Bucket) {
```
does the SDK not do such validation already?

benzekrimaha replied:
This explicit check is there to preserve the previous behavior (early failure with our InvalidRequest error) and avoid sending obviously invalid requests

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,889 @@
+     * @param {object} params - S3 createMultipartUpload params
+     * @param {function} callback
+     */
+    createMultipartUpload(params, callback) {
```
do we actually need the multipart upload for GCP?

seems to me it was not "reimplemented" with SDK v2, so
- either it worked with the "official" SDK : not need to impleemnt some kind of emulation
- or it was not needed : not sure we actually need it (may be handled by the MultiBackend layer), and we should certainly not implement it in the middle of the migration

benzekrimaha replied:
Actually it's not reimplemented , we had these methods in the gcpApis directory, and we used to have them calling the Object.assign , while doing the migration I actually opted for having everything in the service directly as we no longer rely on the service definition like before, the implementation remains unchanged. I can revert back to using the gcpApis like before

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -26,47 +44,67 @@ class AwsClient {
+                    error: err,
+                    method: 'AwsClient.setup',
+                });
+                return done(err);
```
never call `done` from within a try/catch block : it can cause the callback to be invoke multiple times if it throws...

(may not happen all the time in this case, since this is a callback in catch : but we don't _know_ it will not happen, the callback could be called synchronously)

francoisferrand replied:
as disucssed in backbeat, it also relates to the "inverted" logic ("apply region" function than actual - which is less readable) introduced by the mixing of async and callback code with conditional path in between...

This would be solved cleanly by fully moving to async, though probably too much for a single step.

To keep the proper sequence (and reduce diff), maybe you can use async.waterflow:

```
async.waterflow([
    cb => {
        usEast1Client.send(new GetBucketLocationCommand({ Bucket: this._awsBucketName }))
            .then(res => cb(nil, res.LocationConstraint))
            .catch(err => {
                if (err.name === 'AuthorizationHeaderMalformed') {
                    return cb(nil, err.region);
                }
                return cb(err)
            });
    },
    (region, cb) => {
        if (!region) {
            return cb()
        }
    }
], err => {
    done(err)
})
```

or use callbackify:
```
return callbackify(async () => {
    let region
    try {
        region = usEast1Client.send(new GetBucketLocationCommand({ Bucket: this._awsBucketName }))
    } catch(err) {
        if (err.code !== 'AuthorizationHeaderMalformed') {
            this._logger.error('error during setup', {
                error: err,
                method: 'AwsClient.setup',
            });
            throw err;
        }
    }

    if (!region) {
        return;
    }

    ...
    this._client = new S3Client(newConfig);
});
```

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -563,60 +596,54 @@ class AwsClient {
+                .customizeDescription('Error returned from ' +
+                `${this.type}: ${err.message}`));
         });
     }
```
missing new line

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -90,35 +128,39 @@ class AwsClient {
         const log = createLogger(reqUids);
 
-        const putCb = (err, data) => {
+        const putCb = async (err, data) => {
```
this function is not `async`

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -90,35 +128,39 @@ class AwsClient {
                 return callback(errorInstances.ServiceUnavailable
                     .customizeDescription('Error returned from ' +
-                  `${this.type}: ${err.message}`),
+                `${this.type}: ${err.message}`),
```
bad indent

SylvainSenechal reviewed (changes requested):
test

benzekrimaha committed:
last fixes

benzekrimaha committed:
stream handling

benzekrimaha committed:
stream handling

benzekrimaha committed:
getbucket and putObjectCopy reimplem

SylvainSenechal commented:
About the stream : 
- Do we confirm we want to keep _normalizeSdkStream even though we probably don't need to handle the "fromWeb" part, and the body is likely always readable ?
- the stream has 3 methods : abort/destroy/createReadStream. destroy is used for sure, but I feel like abort is useless ? And createReadStream, I thought it was use in the datawrapper.js file, but I don't see it anymore, is it used somewhere else ?
- The part about rawBody !== stream is a bit confusing, and I think it wouldn't be necessary if we assumed that the body was always Readable
- nit: Is the boolean "finished" usefull, I think destroying stream is idempotent

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +236,139 @@ class AwsClient {
                     logLevel = 'info';
                     retError = errors.LocationNotFound;
-                } else {
+                }
```
nit: incorrect line wrap

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +236,139 @@ class AwsClient {
+            .then(data => {
+                const rawBody = data.Body;
+                const stream = this._normalizeSdkStream(rawBody);
+                if (!stream || typeof stream.on !== 'function') {
```
no reason to check that `stream.on` is defined : this is the whole purpose of `_normalizeSdkStream`, and it should return `null` if it does not know how to create a stream from the input...

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +236,139 @@ class AwsClient {
+                stream.abort = abortRequest;
+                stream.destroy = err => abortRequest(err);
+                stream.createReadStream = () => stream;
+
```
somehow here we are recreating the whole stream : instead of modifying the "normalized" object returned by AWS SDK, should we not just create the appropriate "stream" (as we need for our use-case) which will return AWS' stream in its `createReadStream` method?

```suggestion
const stream = {
    abort: abortRequest,
    destroy: abortRequest,
    createReadStream: () => data.Body,
};
```

francoisferrand commented on lib/storage/data/external/AwsClient.js:
```diff
@@ -140,78 +236,139 @@ class AwsClient {
+                        rawBody.destroy?.(err);
+                    }
+                    try {
+                        originalDestroy(err);
```
should destroy be called on `abort()`? should it not instead be called only on `destroy()` ?
(not sure what the semantics of JS streams actually are)

francoisferrand commented on lib/storage/data/external/GcpClient.js:
```diff
@@ -298,6 +304,42 @@ class GcpClient extends AwsClient {
         });
     }
+
+    listObjects(params, callback) {
```
why do we need these extra methods?
the required methods should be already implemented in `AwsClient`...

francoisferrand commented on tests/unit/s3routes/routesUtils/responseStreamData.spec.js:
```diff
@@ -158,36 +160,49 @@ describe('routesUtils.responseStreamData', () => {
+                }, 1000),
+                isclosed: true,
+            }, undefined, logger.newRequestLogger());
+        } catch (err) {
```
error should be handled through some callback : if we catch an exception here, we may have the bug in production code as well....
--> find where the exception is generated
--> handle it properly, sending the response through appropriate callback (or tweak tests/mocks to properly receive it, and fail the test with `done(err)` accordingly)
--> remove the try/catch and useless `once`

francoisferrand commented on tests/unit/storage/data/DummyService.js:
```diff
@@ -81,75 +93,161 @@ class AzureDummyContainerClient {
 class DummyService {
     constructor(config = {}) {
         this.versioning = config.versioning;
+        this.putObjectAsync = promisify(this.putObject.bind(this));
+        this.copyObjectAsync = promisify(this.copyObject.bind(this));
+        this.headObjectAsync = promisify(this.headObject.bind(this));
+        this.putObjectTaggingAsync = promisify(this.putObjectTagging.bind(this));
+        this.deleteObjectTaggingAsync = promisify(this.deleteObjectTagging.bind(this));
+        this.completeMultipartUploadAsync = promisify(this.completeMultipartUpload.bind(this));
+        this.deleteObjectAsync = promisify(this.deleteObject.bind(this));
+        this.createMultipartUploadAsync = promisify(this.createMultipartUpload.bind(this));
+        this.uploadPartAsync = promisify(this.uploadPart.bind(this));
+        this.listPartsAsync = promisify(this.listParts.bind(this));
+        this.abortMultipartUploadAsync = promisify(this.abortMultipartUpload.bind(this));
+        
```
AFAIU, these methods "emulate" the AWS sdk : since we now use them through `send()` (in AWSClient, etc...), no point in promisify (esp. since every method was moved, so the whole file is changed) and best to async' every function

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,619 @@
+                if (metaVersionId) {
+                    result.MetaVersionId = metaVersionId;
+                }
+                return callback && callback(null, result);
```
callback must be set, this is the semantic of this function...
(same later)

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,619 @@
-    },
+    }
+
+    headBucket(params, callback) {
```
not really needed? this is used only in `GcpClient.healthcheck`

```
    healthcheck(location, callback) {
        const checkBucketHealth = (bucket, cb) => {
            let bucketResp;
            this._client.headBucket({ Bucket: bucket }, err => {
```

francoisferrand replied:
need to check other of these "new" API here, and add them only if really needed

francoisferrand commented on lib/storage/data/external/GCP/GcpApis/putObject.js:
```diff
@@ -1,11 +1,43 @@
+
+const { getPutTagsMetadata, attachHeaderCaptureMiddleware } = require('../GcpUtils');
 
 function putObject(params, callback) {
```
where are these functions used? directly in cloudserver?

SylvainSenechal committed:
aws sdk v3 migration : reduce get stream complexity

ISSUE: ARSN-546

SylvainSenechal committed:
bump version to 8.3.0 preview 1

ISSUE: ARSN-546

bert-e commented:
# Wrong destination branch



This pull request targets branch `development/8.4` but this branch doesn't exist.

It is very likely that the development branch was archived before the pull
request got merged.

Please edit the pull request to target a valid development branch.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-514 contains:


* `8.4.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.3.0`


Please check the `Fix Version/s` of ARSN-514, or the target
branch of this pull request.

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,621 @@
+ *   - MPU (multipart upload) is emulated using GCP's compose/copy APIs.
+ *   - All S3-compatible methods are available, with overrides for GCP quirks.
+ */
+class GcpClient extends S3Client {
```
should be rename GcpService or GCP to avoid having 2 times the same type, which can only create confusion

francoisferrand commented on lib/storage/data/external/GcpClient.js:
```diff
@@ -298,6 +306,42 @@ class GcpClient extends AwsClient {
         });
     }
+
+    listObjects(params, callback) {
+        return this.send(new ListObjectsCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    putObject(params, callback) {
+        return this.send(new PutObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    getObject(params, callback) {
+        return this.send(new GetObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    deleteObject(params, callback) {
+        return this.send(new DeleteObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    copyObject(params, callback) {
+        return this.send(new CopyObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
+
+    headObject(params, callback) {
+        return this.send(new HeadObjectCommand(params))
+            .then(data => callback?.(null, data))
+            .catch(err => callback?.(err));
+    }
```
if these functions are used/needed (i.e. part of the abstraction), they should be added to AwsClient ;
if not, they should not be added (at all)

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -1,141 +1,621 @@
+     * @param {object} params - S3 upload params
+     * @param {function} callback
+     */
+    upload(params, callback) {
```
all (or most of) these new functions are not used : this class is used as _client in `GcpClient` (.js), where only the following API are called:
`headBucket`, `creatMultipoartUpdload`, `completeMultipartUpload`, `uploadPart`, `uploadPartCopy` and `abortMultipartUpload` (in addition to the `send` method inherited from S3Client)

francoisferrand replied:
actually it is matching: the functions renamed when they are exported, in `index.js`...

francoisferrand reviewed (approved):
Added a few comments about GCP, I think we are adding lots of dead code: should be cleaned up (but maybe in a followup)

- `GcpClient` class in GcpService.js should be renamed to avoid confusion/duplicated type name → https://github.com/scality/Arsenal/pull/2482#discussion_r2669097878
- It seems the functions in GcpApis are not used (they are 'added' in GcpService's class, but not called... some leftover before code was moved "inside" that class maybe, long before this PR ?) → https://github.com/scality/Arsenal/pull/2482#discussion_r2653193810
- the sdk v2 functions added in GcpClient.js seem weird : if we need to keep the old syntax, should be in AwsClient (esp. since the code there is not Gcp-specific) → https://github.com/scality/Arsenal/pull/2482#discussion_r2669165512 and https://github.com/scality/Arsenal/pull/2482#discussion_r2653143963
- most sdk v2 functions added in GcpService.js seem not to be called → https://github.com/scality/Arsenal/pull/2482#discussion_r2669173602 and https://github.com/scality/Arsenal/pull/2482#discussion_r2653187442
- code for MPU looks fishy: in GcpClient.js:121 we call ` this._client.createMultipartUpload` but this call was not migrated to new SDK nor implemented in GcpService.js.... and thus I think MPU are not tested with GCP
- DummyService.js should get cleaned' up → https://github.com/scality/Arsenal/pull/2482#discussion_r2653172407

francoisferrand commented:
/bypass_author_approval
